File: ios/Packages/OpenUILang/Sources/OpenUILang/Lexer.swift
import Foundation

/// Token kinds mirroring lang-core `parser/tokens` (numeric T.* constants);
/// spec/openui-lang.md §3.1 tokens.
enum Token: Equatable {
    case newline
    case lparen, rparen, lbrack, rbrack, lbrace, rbrace
    case comma, colon
    case equals, eqeq, noteq
    case not
    case greater, greaterEq, less, lessEq
    case and, or
    case dot, question
    case plus, minus, star, slash, percent
    case str(String)
    case num(Double)
    case ident(String)
    case type(String)
    case stateVar(String)   // value includes the leading `$`
    case builtin(String)    // value excludes the leading `@`
    case trueTok, falseTok, nullTok
    case eof

    /// Token kinds that count as "values" for the minus disambiguation.
    var isValueToken: Bool {
        switch self {
        case .num, .str, .ident, .type, .rparen, .rbrack,
            .trueTok, .falseTok, .nullTok, .stateVar, .builtin:
            return true
        default:
            return false
        }
    }
}

/// Direct port of lang-core `parser/lexer.js` `tokenize`
/// (spec/openui-lang.md §3 lexical grammar; strings §3.2–3.3, numbers §3.4).
func tokenize(_ src: [Character]) -> [Token] {
    var tokens: [Token] = []
    var i = 0
    let n = src.count

    func isDigit(_ c: Character) -> Bool { c >= "0" && c <= "9" }
    func isAlpha(_ c: Character) -> Bool {
        (c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c == "_"
    }
    func isWordChar(_ c: Character) -> Bool { isAlpha(c) || isDigit(c) }

    while i < n {
        // Skip horizontal whitespace (not newlines — they're significant)
        while i < n && (src[i] == " " || src[i] == "\t" || src[i] == "\r") { i += 1 }
        if i >= n { break }
        let c = src[i]

        if c == "\n" { tokens.append(.newline); i += 1; continue }
        if c == "(" { tokens.append(.lparen); i += 1; continue }
        if c == ")" { tokens.append(.rparen); i += 1; continue }
        if c == "[" { tokens.append(.lbrack); i += 1; continue }
        if c == "]" { tokens.append(.rbrack); i += 1; continue }
        if c == "{" { tokens.append(.lbrace); i += 1; continue }
        if c == "}" { tokens.append(.rbrace); i += 1; continue }
        if c == "," { tokens.append(.comma); i += 1; continue }
        if c == ":" { tokens.append(.colon); i += 1; continue }

        if c == "=" {
            if i + 1 < n && src[i + 1] == "=" { tokens.append(.eqeq); i += 2 }
            else { tokens.append(.equals); i += 1 }
            continue
        }
        if c == "!" {
            if i + 1 < n && src[i + 1] == "=" { tokens.append(.noteq); i += 2 }
            else { tokens.append(.not); i += 1 }
            continue
        }
        if c == ">" {
            if i + 1 < n && src[i + 1] == "=" { tokens.append(.greaterEq); i += 2 }
            else { tokens.append(.greater); i += 1 }
            continue
        }
        if c == "<" {
            if i + 1 < n && src[i + 1] == "=" { tokens.append(.lessEq); i += 2 }
            else { tokens.append(.less); i += 1 }
            continue
        }
        // A lone `&` lexes as `&&`; a lone `|` lexes as `||`.
        if c == "&" {
            tokens.append(.and)
            i += (i + 1 < n && src[i + 1] == "&") ? 2 : 1
            continue
        }
        if c == "|" {
            tokens.append(.or)
            i += (i + 1 < n && src[i + 1] == "|") ? 2 : 1
            continue
        }
        if c == "." { tokens.append(.dot); i += 1; continue }
        if c == "?" { tokens.append(.question); i += 1; continue }
        if c == "+" { tokens.append(.plus); i += 1; continue }
        if c == "*" { tokens.append(.star); i += 1; continue }
        if c == "/" { tokens.append(.slash); i += 1; continue }
        if c == "%" { tokens.append(.percent); i += 1; continue }

        // ── String literal: "..." ─────────────────────────────────────────
        if c == "\"" {
            let start = i
            i += 1
            var isClosed = false
            while i < n {
                if src[i] == "\\" {
                    i += 2 // skip backslash and the escaped character
                } else if src[i] == "\"" {
                    i += 1 // include the closing quote
                    isClosed = true
                    break
                } else {
                    i += 1
                }
            }
            let end = min(i, n)
            let raw = Array(src[start..<end])
            tokens.append(.str(parseDoubleQuotedString(raw: raw, isClosed: isClosed)))
            continue
        }

        // ── String literal: '...' (single quotes) ─────────────────────────
        if c == "'" {
            i += 1
            var result = ""
            while i < n {
                if src[i] == "\\" {
                    i += 1
                    if i < n {
                        let esc = src[i]
                        if esc == "'" { result.append("'") }
                        else if esc == "\\" { result.append("\\") }
                        else if esc == "n" { result.append("\n") }
                        else if esc == "t" { result.append("\t") }
                        else { result.append(esc) } // pass through other escaped chars
                        i += 1
                    }
                } else if src[i] == "'" {
                    i += 1
                    break
                } else {
                    result.append(src[i])
                    i += 1
                }
            }
            tokens.append(.str(result))
            continue
        }

        // ── Minus: negative number literal or subtraction operator ────────
        var startNumber = false
        if c == "-" {
            let afterValue = tokens.last?.isValueToken ?? false
            if !afterValue && i + 1 < n && isDigit(src[i + 1]) {
                startNumber = true // negative number literal
            } else {
                tokens.append(.minus)
                i += 1
                continue
            }
        }

        // ── Number literal: 42, -3, 1.5, 1e10 ────────────────────────────
        if isDigit(c) || startNumber {
            let start = i
            if src[i] == "-" { i += 1 }
            while i < n && isDigit(src[i]) { i += 1 }
            if i < n && src[i] == "." && i + 1 < n && isDigit(src[i + 1]) {
                i += 1
                while i < n && isDigit(src[i]) { i += 1 }
            }
            if i < n && (src[i] == "e" || src[i] == "E") {
                i += 1
                if i < n && (src[i] == "+" || src[i] == "-") { i += 1 }
                while i < n && isDigit(src[i]) { i += 1 }
            }
            let slice = String(src[start..<i])
            // JS `+slice` semantics: a trailing exponent marker yields NaN.
            tokens.append(.num(Double(slice) ?? .nan))
            continue
        }

        // ── State variable: $identifier ───────────────────────────────────
        if c == "$" && i + 1 < n && isAlpha(src[i + 1]) {
            let start = i
            i += 1
            while i < n && isWordChar(src[i]) { i += 1 }
            tokens.append(.stateVar(String(src[start..<i])))
            continue
        }

        // ── Keyword or identifier ─────────────────────────────────────────
        if isAlpha(c) {
            let start = i
            while i < n && isWordChar(src[i]) { i += 1 }
            let word = String(src[start..<i])
            if word == "true" { tokens.append(.trueTok); continue }
            if word == "false" { tokens.append(.falseTok); continue }
            if word == "null" { tokens.append(.nullTok); continue }
            if c >= "A" && c <= "Z" {
                tokens.append(.type(word))
            } else {
                tokens.append(.ident(word))
            }
            continue
        }

        // ── Builtin call: @identifier ─────────────────────────────────────
        if c == "@" && i + 1 < n && isAlpha(src[i + 1]) {
            i += 1
            let start = i
            while i < n && isWordChar(src[i]) { i += 1 }
            tokens.append(.builtin(String(src[start..<i])))
            continue
        }

        i += 1 // skip any other character (e.g. #, emojis)
    }
    tokens.append(.eof)
    return tokens
}

/// Reproduces the double-quote branch: hand the raw slice (with quotes; a
/// closing quote appended if unclosed) to a JSON string parser; on failure
/// fall back to the raw text with the boundary quotes stripped and no
/// unescaping at all.
func parseDoubleQuotedString(raw: [Character], isClosed: Bool) -> String {
    var candidate = raw
    if !isClosed { candidate.append("\"") }
    if let parsed = parseJSONStringLiteral(candidate) {
        return parsed
    }
    // Fallback: rawString.replace(/^"|"$/g, "") — strip one leading and one
    // trailing quote character (which may be the same character).
    var stripped = raw
    if stripped.first == "\"" { stripped.removeFirst() }
    if stripped.last == "\"" { stripped.removeLast() }
    return String(stripped)
}

/// Strict JSON string literal parser (RFC 8259), matching `JSON.parse` on a
/// single string token. Returns nil on any invalid escape, control character,
/// or malformed shape.
func parseJSONStringLiteral(_ chars: [Character]) -> String? {
    guard chars.count >= 2, chars.first == "\"", chars.last == "\"" else { return nil }
    var out = ""
    var i = 1
    let end = chars.count - 1
    var pendingHighSurrogate: UInt32? = nil

    func flushSurrogate() {
        if let high = pendingHighSurrogate {
            // KNOWN-DEVIATION (README.md #2): JS preserves lone UTF-16
            // surrogates; Swift String cannot — substitute U+FFFD.
            out.append("\u{FFFD}")
            _ = high
            pendingHighSurrogate = nil
        }
    }

    while i < end {
        let c = chars[i]
        if c == "\\" {
            i += 1
            guard i < end else { return nil }
            let e = chars[i]
            switch e {
            case "\"": flushSurrogate(); out.append("\"")
            case "\\": flushSurrogate(); out.append("\\")
            case "/": flushSurrogate(); out.append("/")
            case "b": flushSurrogate(); out.append("\u{08}")
            case "f": flushSurrogate(); out.append("\u{0C}")
            case "n": flushSurrogate(); out.append("\n")
            case "r": flushSurrogate(); out.append("\r")
            case "t": flushSurrogate(); out.append("\t")
            case "u":
                var value: UInt32 = 0
                for k in 1...4 {
                    guard i + k < chars.count, let d = chars[i + k].hexDigitValue else { return nil }
                    value = value * 16 + UInt32(d)
                }
                i += 4
                if value >= 0xD800 && value <= 0xDBFF {
                    flushSurrogate()
                    pendingHighSurrogate = value
                } else if value >= 0xDC00 && value <= 0xDFFF {
                    if let high = pendingHighSurrogate {
                        let combined = 0x10000 + ((high - 0xD800) << 10) + (value - 0xDC00)
                        if let scalar = Unicode.Scalar(combined) {
                            out.unicodeScalars.append(scalar)
                        } else {
                            out.append("\u{FFFD}")
                        }
                        pendingHighSurrogate = nil
                    } else {
                        out.append("\u{FFFD}") // lone low surrogate
                    }
                } else {
                    flushSurrogate()
                    if let scalar = Unicode.Scalar(value) {
                        out.unicodeScalars.append(scalar)
                    } else {
                        out.append("\u{FFFD}")
                    }
                }
            default:
                return nil // invalid escape → whole-string raw fallback
            }
            i += 1
        } else {
            flushSurrogate()
            // JSON.parse rejects unescaped control characters (< U+0020).
            for scalar in c.unicodeScalars where scalar.value < 0x20 {
                _ = scalar
                return nil
            }
            if c == "\"" { return nil } // interior unescaped quote (defensive)
            out.append(c)
            i += 1
        }
    }
    flushSurrogate()
    return out
}

File: android/openui-lang/src/main/kotlin/dev/appless/openuilang/Lexer.kt
package dev.appless.openuilang

/** Token kinds mirroring lang-core `parser/tokens` (spec/openui-lang.md §3.1). */
internal enum class TokType {
    NEWLINE,
    LPAREN, RPAREN, LBRACK, RBRACK, LBRACE, RBRACE,
    COMMA, COLON,
    EQUALS, EQEQ, NOTEQ,
    NOT,
    GREATER, GREATER_EQ, LESS, LESS_EQ,
    AND, OR,
    DOT, QUESTION,
    PLUS, MINUS, STAR, SLASH, PERCENT,
    STR, NUM,
    IDENT, TYPE,
    STATE_VAR, // value INCLUDES the leading `$`
    BUILTIN, // value EXCLUDES the leading `@`
    TRUE, FALSE, NULL,
    EOF,
}

/**
 * A lexed token. `s` carries the text value for STR/IDENT/TYPE/STATE_VAR/
 * BUILTIN, `n` the numeric value for NUM.
 */
internal class Token(val t: TokType, val s: String = "", val n: Double = 0.0) {
    /** Token kinds that count as "values" for the minus disambiguation (§3.4). */
    val isValueToken: Boolean
        get() = when (t) {
            TokType.NUM, TokType.STR, TokType.IDENT, TokType.TYPE, TokType.RPAREN,
            TokType.RBRACK, TokType.TRUE, TokType.FALSE, TokType.NULL,
            TokType.STATE_VAR, TokType.BUILTIN,
            -> true

            else -> false
        }

    override fun toString(): String = "$t(${if (t == TokType.NUM) n.toString() else s})"
}

private fun isDigit(c: Char): Boolean = c in '0'..'9'

private fun isAlpha(c: Char): Boolean = (c in 'a'..'z') || (c in 'A'..'Z') || c == '_'

private fun isWordChar(c: Char): Boolean = isAlpha(c) || isDigit(c)

/**
 * Direct port of lang-core `parser/lexer.js` `tokenize`
 * (spec/openui-lang.md §3 lexical grammar; strings §3.2–3.3, numbers §3.4).
 *
 * Scans the Kotlin `String` by `Char`, which IS a UTF-16 code unit — exactly
 * what JS `src[i]` indexes. No conversion layer is needed (the Swift port has
 * to scan `[UInt16]` because Swift `String` iterates grapheme clusters, which
 * breaks on CRLF and on combining marks).
 *
 * Range checks are code-unit exact by construction: e.g. U+212A KELVIN SIGN
 * (canonically "K") is a single code unit and does NOT satisfy `c in 'A'..'Z'`,
 * matching the JS lexer.
 */
internal fun tokenize(src: String): List<Token> {
    val tokens = ArrayList<Token>()
    var i = 0
    val n = src.length

    while (i < n) {
        // Horizontal whitespace only — `\n` is a significant token. A lone `\r`
        // (and the `\r` of a CRLF pair) is horizontal whitespace (spec §2).
        while (i < n && (src[i] == ' ' || src[i] == '\t' || src[i] == '\r')) i++
        if (i >= n) break
        val c = src[i]

        when (c) {
            '\n' -> { tokens.add(Token(TokType.NEWLINE)); i++; continue }
            '(' -> { tokens.add(Token(TokType.LPAREN)); i++; continue }
            ')' -> { tokens.add(Token(TokType.RPAREN)); i++; continue }
            '[' -> { tokens.add(Token(TokType.LBRACK)); i++; continue }
            ']' -> { tokens.add(Token(TokType.RBRACK)); i++; continue }
            '{' -> { tokens.add(Token(TokType.LBRACE)); i++; continue }
            '}' -> { tokens.add(Token(TokType.RBRACE)); i++; continue }
            ',' -> { tokens.add(Token(TokType.COMMA)); i++; continue }
            ':' -> { tokens.add(Token(TokType.COLON)); i++; continue }
            '.' -> { tokens.add(Token(TokType.DOT)); i++; continue }
            '?' -> { tokens.add(Token(TokType.QUESTION)); i++; continue }
            '+' -> { tokens.add(Token(TokType.PLUS)); i++; continue }
            '*' -> { tokens.add(Token(TokType.STAR)); i++; continue }
            '/' -> { tokens.add(Token(TokType.SLASH)); i++; continue }
            '%' -> { tokens.add(Token(TokType.PERCENT)); i++; continue }
            else -> Unit
        }

        if (c == '=') {
            if (i + 1 < n && src[i + 1] == '=') { tokens.add(Token(TokType.EQEQ)); i += 2 }
            else { tokens.add(Token(TokType.EQUALS)); i++ }
            continue
        }
        if (c == '!') {
            if (i + 1 < n && src[i + 1] == '=') { tokens.add(Token(TokType.NOTEQ)); i += 2 }
            else { tokens.add(Token(TokType.NOT)); i++ }
            continue
        }
        if (c == '>') {
            if (i + 1 < n && src[i + 1] == '=') { tokens.add(Token(TokType.GREATER_EQ)); i += 2 }
            else { tokens.add(Token(TokType.GREATER)); i++ }
            continue
        }
        if (c == '<') {
            if (i + 1 < n && src[i + 1] == '=') { tokens.add(Token(TokType.LESS_EQ)); i += 2 }
            else { tokens.add(Token(TokType.LESS)); i++ }
            continue
        }
        // A single `&` lexes as `&&`; a single `|` lexes as `||` (spec §3.1).
        if (c == '&') {
            tokens.add(Token(TokType.AND))
            i += if (i + 1 < n && src[i + 1] == '&') 2 else 1
            continue
        }
        if (c == '|') {
            tokens.add(Token(TokType.OR))
            i += if (i + 1 < n && src[i + 1] == '|') 2 else 1
            continue
        }

        // ── String literal: "..." ────────────────────────────────────────────
        if (c == '"') {
            val start = i
            i++
            var isClosed = false
            while (i < n) {
                when {
                    src[i] == '\\' -> i += 2 // skip backslash and the escaped char
                    src[i] == '"' -> { i++; isClosed = true }
                    else -> i++
                }
                if (isClosed) break
            }
            val end = if (i < n) i else n
            val raw = src.substring(start, end)
            tokens.add(Token(TokType.STR, s = parseDoubleQuotedString(raw, isClosed)))
            continue
        }

        // ── String literal: '...' (single quotes) ────────────────────────────
        if (c == '\'') {
            i++
            val result = StringBuilder()
            while (i < n) {
                if (src[i] == '\\') {
                    i++
                    if (i < n) {
                        when (val esc = src[i]) {
                            '\'' -> result.append('\'')
                            '\\' -> result.append('\\')
                            'n' -> result.append('\n')
                            't' -> result.append('\t')
                            else -> result.append(esc) // pass through
                        }
                        i++
                    }
                } else if (src[i] == '\'') {
                    i++
                    break
                } else {
                    result.append(src[i])
                    i++
                }
            }
            tokens.add(Token(TokType.STR, s = result.toString()))
            continue
        }

        // ── Minus: negative-number prefix or subtraction operator ────────────
        var startNumber = false
        if (c == '-') {
            val afterValue = tokens.isNotEmpty() && tokens[tokens.size - 1].isValueToken
            if (!afterValue && i + 1 < n && isDigit(src[i + 1])) {
                startNumber = true
            } else {
                tokens.add(Token(TokType.MINUS))
                i++
                continue
            }
        }

        // ── Number literal: 42, -3, 1.5, 1e10 ───────────────────────────────
        if (isDigit(c) || startNumber) {
            val start = i
            if (src[i] == '-') i++
            while (i < n && isDigit(src[i])) i++
            // The decimal point is consumed only if a digit follows it.
            if (i < n && src[i] == '.' && i + 1 < n && isDigit(src[i + 1])) {
                i++
                while (i < n && isDigit(src[i])) i++
            }
            if (i < n && (src[i] == 'e' || src[i] == 'E')) {
                i++
                if (i < n && (src[i] == '+' || src[i] == '-')) i++
                while (i < n && isDigit(src[i])) i++
            }
            // JS `+slice` semantics. The slice can only contain
            // [0-9.eE+-], so `toDoubleOrNull` agrees with `Number()` — a
            // trailing exponent marker ("1e") yields NaN in both.
            tokens.add(Token(TokType.NUM, n = src.substring(start, i).toDoubleOrNull() ?: Double.NaN))
            continue
        }

        // ── State variable: $identifier ─────────────────────────────────────
        if (c == '$' && i + 1 < n && isAlpha(src[i + 1])) {
            val start = i
            i++
            while (i < n && isWordChar(src[i])) i++
            tokens.add(Token(TokType.STATE_VAR, s = src.substring(start, i)))
            continue
        }

        // ── Keyword or identifier ───────────────────────────────────────────
        if (isAlpha(c)) {
            val start = i
            while (i < n && isWordChar(src[i])) i++
            when (val word = src.substring(start, i)) {
                "true" -> tokens.add(Token(TokType.TRUE))
                "false" -> tokens.add(Token(TokType.FALSE))
                "null" -> tokens.add(Token(TokType.NULL))
                else -> tokens.add(
                    Token(if (c in 'A'..'Z') TokType.TYPE else TokType.IDENT, s = word)
                )
            }
            continue
        }

        // ── Builtin call: @identifier ───────────────────────────────────────
        if (c == '@' && i + 1 < n && isAlpha(src[i + 1])) {
            i++
            val start = i
            while (i < n && isWordChar(src[i])) i++
            tokens.add(Token(TokType.BUILTIN, s = src.substring(start, i)))
            continue
        }

        i++ // silently skip any other code unit (#, emoji surrogates, ;, …)
    }
    tokens.add(Token(TokType.EOF))
    return tokens
}

/**
 * Reproduces the double-quote branch (spec §3.2): hand the raw slice (with
 * quotes; a closing quote appended if unclosed) to a JSON string parser; on
 * ANY failure fall back to the raw text with the boundary quotes stripped and
 * no unescaping at all.
 */
internal fun parseDoubleQuotedString(raw: String, isClosed: Boolean): String {
    val candidate = if (isClosed) raw else raw + "\""
    parseJsonStringLiteral(candidate)?.let { return it }
    // Fallback: rawString.replace(/^"|"$/g, "") — strip one leading and one
    // trailing quote code unit (which may be the same unit).
    var lo = 0
    var hi = raw.length
    if (lo < hi && raw[lo] == '"') lo++
    if (hi > lo && raw[hi - 1] == '"') hi--
    return raw.substring(lo, hi)
}

/**
 * Strict JSON string-literal parser (RFC 8259) over UTF-16 code units,
 * matching `JSON.parse` on a single string token. Returns `null` on any
 * invalid escape, control character, or malformed shape — which is what makes
 * the whole-string raw fallback fire.
 *
 * DELIBERATE DIVERGENCE FROM THE SWIFT PORT (its KNOWN-DEVIATION #2): a `\uXXXX`
 * escape appends its code unit verbatim, so an UNPAIRED surrogate survives
 * exactly as it does in JS. Kotlin `String` can hold a lone surrogate; Swift
 * `String` cannot and substitutes U+FFFD there.
 */
internal fun parseJsonStringLiteral(text: String): String? {
    if (text.length < 2 || text[0] != '"' || text[text.length - 1] != '"') return null
    val out = StringBuilder(text.length)
    var i = 1
    val end = text.length - 1

    // JSON hex digits are exactly [0-9a-fA-F] — code-unit exact (JS rejects
    // e.g. fullwidth digits, which `Character.digit` would accept).
    fun hexDigit(u: Char): Int = when (u) {
        in '0'..'9' -> u - '0'
        in 'a'..'f' -> u - 'a' + 10
        in 'A'..'F' -> u - 'A' + 10
        else -> -1
    }

    while (i < end) {
        val c = text[i]
        if (c == '\\') {
            i++
            if (i >= end) return null
            when (text[i]) {
                '"' -> out.append('"')
                '\\' -> out.append('\\')
                '/' -> out.append('/')
                'b' -> out.append('\b')
                'f' -> out.append('\u000C')
                'n' -> out.append('\n')
                'r' -> out.append('\r')
                't' -> out.append('\t')
                'u' -> {
                    var value = 0
                    for (k in 1..4) {
                        if (i + k >= text.length) return null
                        val d = hexDigit(text[i + k])
                        if (d < 0) return null
                        value = value * 16 + d
                    }
                    i += 4
                    out.append(value.toChar())
                }

                else -> return null // invalid escape → whole-string raw fallback
            }
            i++
        } else {
            if (c < ' ') return null // JSON.parse rejects raw control chars
            if (c == '"') return null // interior unescaped quote (defensive)
            out.append(c)
            i++
        }
    }
    return out.toString()
}
